src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,424 @@
+{ inf := λ I J, ⟨I ⊓ J, I.prop.inf J.prop⟩ }
+
+instance : has_Inf (homogeneous_ideal 𝒜) :=
+{ Inf := λ ℐ, ⟨Inf (coe '' ℐ), ideal.is_homogeneous.Inf $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_sup (homogeneous_ideal 𝒜) :=
+{ sup := λ I J, ⟨I ⊔ J, I.prop.sup J.prop⟩ }
+
+instance : has_Sup (homogeneous_ideal 𝒜) :=
+{ Sup := λ ℐ, ⟨Sup (coe '' ℐ), ideal.is_homogeneous.Sup $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_mul (homogeneous_ideal 𝒜) :=
+{ mul := λ I J, ⟨I * J, I.prop.mul J.prop⟩ }
+
+instance : has_add (homogeneous_ideal 𝒜) := ⟨(⊔)⟩
+
+end operations
+
+section homogeneous_core
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A]
+variables [algebra R A] [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+lemma ideal.homogeneous_core.gc :
+  galois_connection
+    (subtype.val : homogeneous_ideal 𝒜 → ideal A)
+    (λ I, ⟨I.homogeneous_core 𝒜, I.is_homogeneous_homogeneous_core 𝒜⟩ :
+      ideal A → homogeneous_ideal 𝒜) :=
+λ I J, ⟨
+  λ H, show I.1 ≤ ideal.homogeneous_core 𝒜 J, begin
+    rw I.2.homogeneous_core_eq_self,
+    exact ideal.homogeneous_core_is_mono 𝒜 H,
+  end,
+  λ H, le_trans H (ideal.homogeneous_core_le_ideal _ _)⟩
+
+/--There is a galois coinsertion between homogeneous ideals and ideals via
+`(λ I, I.1)` and `ideal.homogeneous_core`-/
+def ideal.homogeneous_core.gi :
+  galois_coinsertion
+    (subtype.val : homogeneous_ideal 𝒜 → ideal A)
+    (λ I, ⟨I.homogeneous_core 𝒜, I.is_homogeneous_homogeneous_core 𝒜⟩ :
+      ideal A → homogeneous_ideal 𝒜) :=
+{ choice := λ I HI, ⟨I, begin
+    have eq : I = I.homogeneous_core 𝒜,
+    refine le_antisymm HI _,
+    apply (ideal.homogeneous_core_le_ideal 𝒜 I),
+    rw eq, apply ideal.is_homogeneous_homogeneous_core,
+  end⟩,
+  gc := ideal.homogeneous_core.gc 𝒜,
+  u_l_le := λ I, by apply ideal.homogeneous_core_le_ideal,
+  choice_eq := λ I H, le_antisymm H (I.homogeneous_core_le_ideal _) }
+
+lemma ideal.homogeneous_core_eq_Sup :
+  I.homogeneous_core 𝒜 = Sup {J : ideal A | J.is_homogeneous 𝒜 ∧ J ≤ I} :=
+begin
+  refine (is_lub.Sup_eq _).symm,
+  apply is_greatest.is_lub,
+  have coe_mono : monotone (coe : {I : ideal A // I.is_homogeneous 𝒜} → ideal A) := λ _ _, id,
+  convert coe_mono.map_is_greatest (ideal.homogeneous_core.gc 𝒜).is_greatest_u using 1,
+  simp only [subtype.coe_image, exists_prop, mem_set_of_eq, subtype.coe_mk],
+end
+
+end homogeneous_core
+
+section homogeneous_hull
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A]
+variables [algebra R A] [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--For any `I : ideal R`, not necessarily homogeneous, there is a homogeneous ideal associated with
+`I` spanned by all homogeneous components of elements in `I`. -/
+def ideal.homogeneous_hull : ideal A :=
+  ideal.span {r : A | ∃ (i : ι) (x : I), (graded_algebra.decompose 𝒜 x i : A) = r}
```

Comment:
This might be a useful lemma:
```lean
lemma homogeneous_hull_eq_supr :
  I.homogeneous_hull 𝒜 = ⨆ i, ideal.span (graded_algebra.proj 𝒜 i '' I) :=
begin
  refine eq.trans _ (submodule.span_Union _), -- todo: `ideal.span_Union` so that we can use `rw`
  apply congr_arg ideal.span _,
  ext1, simp only [set.mem_Union, set.mem_image, mem_set_of_eq, graded_algebra.proj_apply,
    set_like.exists, exists_prop, subtype.coe_mk, set_like.mem_coe],
end
```